contracts/governance/FeeSharingProxy.sol

```diff
@@ -0,0 +1,156 @@
+        require(_token != address(0), "FeeSharingProxy::withdrawFees: invalid address");
+        require(
+            block.timestamp - lastFeeWithdrawalTime >= FEE_WITHDRAWAL_INTERVAL,
+            "FeeSharingProxy::withdrawFees: the last withdrawal was recently"
+        );
+    
+        address loanPoolToken = protocol.underlyingToLoanPool(_token);
+        require(loanPoolToken != address(0), "FeeSharingProxy::withdrawFees: loan token not found");
+
+        uint amount = protocol.withdrawFees(_token, address(this));
+        require(amount > 0, "FeeSharingProxy::withdrawFees: no tokens to withdraw");
+
+        //TODO can be also used - function addLiquidity(IERC20Token _reserveToken, uint256 _amount, uint256 _minReturn)
+        IERC20(_token).approve(loanToken, amount);
+        uint poolTokenAmount = ILoanToken(loanPoolToken).mint(address(this), amount);
+        _writeTokenCheckpoint(loanPoolToken, uint96(poolTokenAmount));
+
+        lastFeeWithdrawalTime = block.timestamp;
+        
+        emit FeeWithdrawn(msg.sender, loanPoolToken, poolTokenAmount);
+    }
+
+    /**
+     * @notice withdraw accumulated fee the message sender
+     * @param _loanPoolToken address of the pool token
+     * @param _maxCheckpoints maximum number of checkpoints to be processed
+     * @param _receiver the receiver of tokens or msg.sender
+     * */
+    function withdraw(address _loanPoolToken, uint32 _maxCheckpoints, address _receiver) public {
+        address user = msg.sender;
+        if (_receiver == address(0)) {
+            _receiver = msg.sender;
+        }
+        uint32 start = processedCheckpoints[user][_loanPoolToken];
+        uint32 nCheckpoints = numTokenCheckpoints[_loanPoolToken];
+        require(start < nCheckpoints, "FeeSharingProxy::withdrawFees: no tokens for a withdrawal");
+
+        if (_maxCheckpoints > MAX_CHECKPOINTS) {
+            _maxCheckpoints = MAX_CHECKPOINTS;
+        }
+        uint32 end = safe32(start + _maxCheckpoints, "FeeSharingProxy::withdraw: checkpoint index exceeds 32 bits");
+        if (end > nCheckpoints) {
+            end = nCheckpoints;
+        }
+        //Withdrawal should only be possible for blocks which were already mined.
+        uint32 lastBlockNumber = tokenCheckpoints[_loanPoolToken][end - 1].blockNumber;
+        if (block.number == lastBlockNumber) {
+            end--;
+        }
+        uint256 amount = 0;
+        uint cachedLockDate = 0;
+        uint96 cachedWeightedStake = 0;
+        for (uint32 i = start; i < end; i++) {
+            Checkpoint storage checkpoint = tokenCheckpoints[_loanPoolToken][i];
+            uint lockDate = staking.timestampToLockDate(checkpoint.timestamp);
+            uint96 weightedStake;
+            if (lockDate == cachedLockDate) {
+                weightedStake = cachedWeightedStake;
+            } else {
+                weightedStake = staking.getPriorWeightedStake(user, checkpoint.blockNumber - 1, checkpoint.timestamp);
```

Comment:
i think here a comment would be good explaining that we use blocknumber -1, because the total voting power was also timestamped for block number -1